### PR TITLE
fix(ec2): ARM-backed bastion hosts try to run x86-based Amazon Linux AMI

### LINKED ISSUE #12279: (ec2): using ARM-based instances in BastionHostLinux results in an invalid instance configuration
Deployment of a stack containing a BastionHostLinux construct backed by an ARM-based instance (e.g., `a1`, `t4g`, etc) will fail, with CloudFormation reporting the following for the EC2 instance resource:
```
't4g.nano' is not a valid instance type for instance 'i-*********' of architecture 'x86_64'. (Service: AmazonEC2; Status Code: 400; Error Code: InvalidInstanceAttributeValue; Request ID: ********; Proxy: null)
```

### Reproduction Steps
The following stack will fail during deployment:
```javascript
const cdk = require('@aws-cdk/core');
const ec2 = require('@aws-cdk/aws-ec2');
class BastionHostReproStack extends cdk.Stack {

  constructor(scope, id, props) {
    super(scope, id, props);

    this.vpc = new ec2.Vpc(this, 'Vpc');
    new ec2.BastionHostLinux(this, 'BastionHost', {
      vpc: this.vpc,
      instanceType: ec2.InstanceType.of(ec2.InstanceClass.T4G, ec2.InstanceSize.NANO),
    });
  }
}

module.exports = { BastionHostReproStack }
```

### What did you expect to happen?
The stack should deploy successfully when using an instanceType on BastionHostLinux that has an InstanceClass with an ARM architecture.

### What actually happened?
CloudFormation will report the following for the backing EC2 instance:
```
't4g.nano' is not a valid instance type for instance 'i-*********' of architecture 'x86_64'. (Service: AmazonEC2; Status Code: 400; Error Code: InvalidInstanceAttributeValue; Request ID: ********; Proxy: null)
```

### Environment

  - **CDK CLI Version  :** 1.80
  - **Framework Version:** 1.80
  - **Node.js Version:** 15.2.1
  - **OS               :** Ubuntu (WSL2)
  - **Language (Version):** TypeScript, JavaScript

### Other
In the constructor of BastionHostLinux, the architecture of the instanceType is assumed to be x86-64, and so the Amazon Linux 2 X86-64 AMI is always used for the backing EC2 instance even if the instanceType uses an instance class that is ARM-based. This explains the error reported by CloudFormation.
Furthermore, only the x86-64 SSM Agent rpm is installed via UserData.

Proposed fix at https://github.com/aws/aws-cdk/pull/12280

--- 

This is :bug: Bug Report

### PR DESCRIPTION
**Pull request by Jacob-Doetsch**
Fixes #12279

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-ec2/lib/instance-types.ts [hunk 1]
```diff
@@ -565,4 +586,21 @@ export class InstanceType {
   public toString(): string {
     return this.instanceTypeIdentifier;
   }
+
+  /**
+   * Returns the CPU architecture for the instance type
+   */
+  private resolveArchitecture(): InstanceArchitecture {
+    const ARM_INSTANCE_CLASSES = new Set<InstanceClass>([
+      InstanceClass.A1, InstanceClass.C6G, InstanceClass.C6GD,
```
Comment (Jacob-Doetsch) [inline packages/@aws-cdk/aws-ec2/lib/instance-types.ts:595]: An alternative to this explicit checking against a set was instead to test the target instanceClass against a regular expression. A regex for testing for Graviton or Graviton2 turns out to be pretty straightforward, e.g.:
```javascript
// matches a1, t4g, c6g, c6gd, c6gn, m6g, m6gd, r6g, r6gd
/^a1|[cmrt][46]g[dn]?$/
```
Thoughts?

File: packages/@aws-cdk/aws-ec2/lib/instance-types.ts [hunk 2]
```diff
@@ -556,7 +571,13 @@ export class InstanceType {
     return new InstanceType(`${instanceClass}.${instanceSize}`);
   }
 
+  /**
+   * The instance's CPU architecture
+   */
+  public readonly architecture: InstanceArchitecture;
```
Comment (Jacob-Doetsch) [inline packages/@aws-cdk/aws-ec2/lib/instance-types.ts:577]: Wasn't sure if it would have been better to forgo a new property and instead just make the `resolveArchitecture` method public. Thoughts?

File: packages/@aws-cdk/aws-ec2/lib/instance-types.ts (hunk 1 above)
Comment (hoegertn) [reply packages/@aws-cdk/aws-ec2/lib/instance-types.ts:595]: I think you could target all digits instead of only 4 and 6. The important part is the `g` after the generation.

Comment (Jacob-Doetsch) [reply packages/@aws-cdk/aws-ec2/lib/instance-types.ts]: That's a great point. I ended up implementing the regex check and used your feedback.

File: packages/@aws-cdk/aws-ec2/lib/bastion-host.ts [hunk 3]
```diff
@@ -176,13 +177,23 @@ export class BastionHostLinux extends Resource implements IInstance {
...
       description: 'Instance ID of the bastion host. Use this to connect via SSM Session Manager',
       value: this.instanceId,
     });
   }
 
+  /**
+   * Install the architecture-appropriate SSM Agent package
+   */
+  private applyUserData(): void {
+    const instanceType = new InstanceType(this.instance.instance.instanceType!);
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-ec2/lib/bastion-host.ts:192]: Seems safer and easier to have `instanceType` passed in here from the constructor.

File: packages/@aws-cdk/aws-ec2/lib/bastion-host.ts [hunk 4]
```diff
@@ -146,14 +146,17 @@ export class BastionHostLinux extends Resource implements IInstance {
...
       vpc: props.vpc,
       availabilityZone: props.availabilityZone,
       securityGroup: props.securityGroup,
       instanceName: props.instanceName ?? 'BastionHost',
-      instanceType: props.instanceType ?? InstanceType.of(InstanceClass.T3, InstanceSize.NANO),
-      machineImage: props.machineImage ?? MachineImage.latestAmazonLinux({ generation: AmazonLinuxGeneration.AMAZON_LINUX_2 }),
+      instanceType,
+      machineImage: props.machineImage ?? MachineImage.latestAmazonLinux({
+        generation: AmazonLinuxGeneration.AMAZON_LINUX_2,
+        cpuType: instanceType.architecture as string as AmazonLinuxCpuType,
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-ec2/lib/bastion-host.ts:158]: This is making me feel uneasy. I know it's a chore but how about an explicit conversion function?

File: packages/@aws-cdk/aws-ec2/lib/instance-types.ts (hunk 2 above)
Comment (rix0rrr) [reply packages/@aws-cdk/aws-ec2/lib/instance-types.ts:597]: I like the current public name. The best of both worlds would be to make `resolveArchitecture` a getter :)

```ts
public get architecture(): InstanceArchitecture {
  // ...body of resolveArchitecture here...
}
```

File: packages/@aws-cdk/aws-ec2/lib/instance-types.ts [hunk 5]
```diff
@@ -585,4 +606,18 @@ export class InstanceType {
...
   }
+
+  /**
+   * Returns the CPU architecture for the instance type
+   */
+  private resolveArchitecture(): InstanceArchitecture {
+    const instanceClass = this.instanceTypeIdentifier.split('.')[0];
+
+    // Graviton-based (a1, t4g, c6g, c6gd, c6gn, m6g, m6gd, r6g, r6gd)
+    if (/^a1|[cmrt][\d]g[dn]?$/.test(instanceClass)) {
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-ec2/lib/instance-types.ts:617]: I like doing it by regex, but I think we should simplify the regex some so we can automatically account for future instance  types (otherwise this will be chasing a moving target).

An ARM instance is identified by either:

* Starts with an `a` for instance family ARM; or
* Has a `g` in the "extra capabilities" for Graviton

Also we're getting dangerously close to generation 10 so I'd like to be safe against double-digit generations.


I'll propose:

```
/^a|^[^.]*\d[^.]*g[^.]*\./
```

In other words: starts with an `a` OR has a `g` somewhere after a digit before the first period.

Probably worth writing down what it's testing for in a comment next to it :)



image

**Review by rix0rrr — DISMISSED** [review]
I like what you're doing. Some small comments.

Something that occurred to me: don't recent Amazon Linux AMIs come with the SSM Agent preinstalled? Why do we even need to `yum install` it at all?

File: packages/@aws-cdk/aws-ec2/lib/bastion-host.ts (hunk 4 above)
Comment (Jacob-Doetsch) [reply packages/@aws-cdk/aws-ec2/lib/bastion-host.ts]: Great suggestion, thank you for the feedback

File: packages/@aws-cdk/aws-ec2/lib/instance-types.ts (hunk 5 above)
Comment (Jacob-Doetsch) [reply packages/@aws-cdk/aws-ec2/lib/instance-types.ts]: Thank you for the feedback, I have attempted to improve the regex and make it more future-proof

**Comment by Jacob-Doetsch** [comment]
@rix0rrr You're right, there is no reason to manually install the SSM Agent in this case since it is preinstalled on Amazon Linux 2. I have removed the UserData stuff.

File: packages/@aws-cdk/aws-ec2/lib/bastion-host.ts (hunk 3 above)
Comment (Jacob-Doetsch) [reply packages/@aws-cdk/aws-ec2/lib/bastion-host.ts]: I confirmed that the SSM Agent comes preinstalled on AL2 for both x86-64 and aarch64, so we have an opportunity to simplify. I have removed the userdata logic.

File: packages/@aws-cdk/aws-ec2/lib/instance-types.ts (hunk 5 above)
Comment (Jacob-Doetsch) [reply packages/@aws-cdk/aws-ec2/lib/instance-types.ts]: I refined it yet again. Now we just match against the whole instanceTypeIdentifier and capture the family-part and capabilities-part and test against those instead.

File: packages/@aws-cdk/aws-ec2/lib/instance-types.ts [hunk 6]
```diff
@@ -585,4 +600,22 @@ export class InstanceType {
   public toString(): string {
     return this.instanceTypeIdentifier;
   }
+
+  /**
+   * The instance's CPU architecture
+   */
+  public get architecture(): InstanceArchitecture {
+    // capture the family, generation, and capabilities portion of the instance type
+    const instanceTypePattern = this.instanceTypeIdentifier.match(/^([a-z]+)(\d{1,2})([a-z]*)\..*$/);
```
Comment (rix0rrr) [inline packages/@aws-cdk/aws-ec2/lib/instance-types.ts:609]: Deal with the possibility of this not matching please. A good rule of thumb when programming with or for other people: "trust nothing" 😄

**Review by rix0rrr — CHANGES_REQUESTED** [review]
Almost there!

**Comment by Jacob-Doetsch** [comment]
@rix0rrr Let me know if I have addressed all your concerns. Thanks!

**Review by rix0rrr — APPROVED** [review]
